backward_paper/runner.py

```diff
@@ -0,0 +1,139 @@
+# -*- coding: utf-8 -*-
+"""
+This script contains a specialization of the Ekomark runner
+"""
+import functools
+import pathlib
+import copy
+import pandas as pd
+import numpy as np
+
+from banana import load_config
+from banana.data import dfdict
+
+from ekomark.benchmark.runner import Runner
+from ekomark import pdfname
+
+from plots import plot_pdf
+
+import lhapdf
+import eko
+
+pkg_path = pathlib.Path(__file__).absolute().parents[0]
+
+
+def rotate_to_pm_basis(log):
+    """Rotate to plus minus basis"""
+    # TODO: this can be improved
+    rot_log = {}
+    if "g" in log:
+        rot_log["g"] = log["g"]
+    for pid in eko.evolution_operator.flavors.quark_names:
+        if pid not in log:
+            continue
+        quark = log[pid]
+        qbar = log[f"{pid}bar"].copy()
+        rot_log[r"${%s}^{+}$" % (pid)] = copy.deepcopy(log[pid])
+        rot_log[r"${%s}^{-}$" % (pid)] = copy.deepcopy(log[pid])
+
+        for column_name, column_data in quark.items():
+            if column_name == "x":
+                continue
+            rot_log[r"${%s}^{+}$" % (pid)][column_name] = (
+                column_data + qbar[column_name]
+            )
+            rot_log[r"${%s}^{-}$" % (pid)][column_name] = (
+                column_data - qbar[column_name]
+            )
+    return rot_log
+
+
+class EkoRunner(Runner):
+    """
+    Specialization of the Ekomark runner.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.banana_cfg = load_config(pkg_path)
+        self.rotate_to_evolution_basis = False
+        self.plot_pdfs = [4, -4]
+        self.sandbox = True
+        self.fig_name = None
+
+    def run_external(self, theory, ocard, pdf):
+
+        if self.external == "inputpdf":
+            # Compare with the initial pdf
+
+            xgrid = ocard["interpolation_xgrid"]
+            q0 = theory["Q0"]
+            ext = {}
+            pdfs = lhapdf.mkPDFs(pdf.set().name)
+            for rep, base_pdf in enumerate(pdfs):
+                tab = {}
+                for x in xgrid:
+                    in_pdf = base_pdf.xfxQ(x, q0)
+                    for pid, val in in_pdf.items():
+                        if pid not in tab:
+                            tab[pid] = []
+                        tab[pid].append(val)
+
+                ext[rep] = tab
+        return ext
+
+    def log(self, theory, ocard, _pdf, me, ext):
+        """Apply PDFs to eko and produce log tables"""
+        log_tabs = {}
+        xgrid = ocard["interpolation_xgrid"]
+        q2s = ocard["Q2grid"]
+
+        rotate_to_evolution = None
+        if self.rotate_to_evolution_basis:
+            rotate_to_evolution = eko.basis_rotation.rotate_flavor_to_evolution.copy()
+
+        pdf_name = _pdf.set().name
+        pdfs = lhapdf.mkPDFs(pdf_name)
+
+        # build table
+        tab = {}
+        tab["x"] = xgrid
+
+        # Loop over pdfs replicas
+        for rep, pdf in enumerate(pdfs):
```

Comment:
- actually: this is quite a general feature - we maybe want to lift this into `banana`, (by improving `get_pdf`, e,g,?)
- this may also affect our `run` calls - the pdf argument could e.g. become a tuple `(set_name, member)` and if it is just a string then cast it to `member=0` (as is currently)

Reply:
- if you want all replicas, we could provide a helper function, that spells out the respective tuples